File: DemoStudyteam/demo-AppInstalled/demo-AppInstalled/ViewController.m
//
//  ViewController.m
//  demo-AppInstalled
//
//  Created by Lam Quoc on 3/4/16.
//  Copyright © 2016 Lam Quoc. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    NSArray *bundles2Check = [NSArray arrayWithObjects: @";com.apple.mobilesafari", @";com.yourcompany.yourselfmadeapp", @";com.blahblah.nonexistent",@";Home.demo-AppInstalled",@";demo-AppInstalled", nil];
    for (NSString *identifier in bundles2Check)
        if (APCheckIfAppInstalled(identifier))
            NSLog(@";App installed: %@", identifier);
        else
            NSLog(@";App not installed: %@", identifier);

}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



// Declaration
BOOL APCheckIfAppInstalled(NSString *bundleIdentifier); // Bundle identifier (eg. com.apple.mobilesafari) used to track apps

// Implementation

BOOL APCheckIfAppInstalled(NSString *bundleIdentifier)
{
    static NSString *const cacheFileName = @";com.apple.mobile.installation.plist";
    NSString *relativeCachePath = [[@";Library" stringByAppendingPathComponent: @";Caches"] stringByAppendingPathComponent: cacheFileName];
    NSDictionary *cacheDict = nil;
    NSString *path = nil;
    // Loop through all possible paths the cache could be in
    for (short i = 0; 1; i++)
    {
        
        switch (i) {
            case 0: // Jailbroken apps will find the cache here; their home directory is /var/mobile
                path = [NSHomeDirectory() stringByAppendingPathComponent: relativeCachePath];
                break;
            case 1: // App Store apps and Simulator will find the cache here; home (/var/mobile/) is 2 directories above sandbox folder
                path = [[NSHomeDirectory() stringByAppendingPathComponent: @";../.."] stringByAppendingPathComponent: relativeCachePath];
                break;
            case 2: // If the app is anywhere else, default to hardcoded /var/mobile/
                path = [@";/var/mobile" stringByAppendingPathComponent: relativeCachePath];
                break;
            default: // Cache not found (loop not broken)
                return NO;
            break; }
        
        BOOL isDir = NO;
        if ([[NSFileManager defaultManager] fileExistsAtPath: path isDirectory: &isDir] && !isDir) // Ensure that file exists
            cacheDict = [NSDictionary dictionaryWithContentsOfFile: path];
        
        if (cacheDict) // If cache is loaded, then break the loop. If the loop is not "broken," it will return NO later (default: case)
            break;
    }
    
    NSDictionary *system = [cacheDict objectForKey: @";System"]; // First check all system (jailbroken) apps
    if ([system objectForKey: bundleIdentifier]) return YES;
    NSDictionary *user = [cacheDict objectForKey: @";User"]; // Then all the user (App Store /var/mobile/Applications) apps
    if ([user objectForKey: bundleIdentifier]) return YES;
    
    // If nothing returned YES already, we'll return NO now
    return NO;
}


@end

File: DemoStudyteam/demo-DownAppInAppStore/demo-DownAppInAppStore/ViewController.swift
//
//  ViewController.swift
//  demo-DownAppInAppStore
//
//  Created by Lam Quoc on 3/4/16.
//  Copyright © 2016 Lam Quoc. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


    @IBAction func link1(sender: UIButton) {
        let iTunesLink = "itms://itunes.apple.com/de/app/x-gift/id839686104?mt=8&uo=4"
        
//        [[UIApplication sharedApplication] openURL:[NSURL URLWithString:iTunesLink]];
        
        
        var url:NSURL   = NSURL()
        
        url = NSURL(string: iTunesLink)!
        
        UIApplication.sharedApplication().openURL(url)
    }
    
    @IBAction func link2(sender: UIButton) {
        //https://itunes.apple.com/us/app/solitaire/id359917414?mt=8
        let iTuneLink = "itms://itunes.apple.com/us/app/solitaire/id359917414?mt=8"
        var url:NSURL = NSURL()
        url = NSURL(string: iTuneLink)!
        UIApplication.sharedApplication().openURL(url)
    }
    
    @IBAction func link3(sender: UIButton) {
        //itms://itunes.apple.com/us/app/charades!-free/id653967729?mt=8
        let iTunelink = "itms://itunes.apple.com/us/app/charades!-free/id653967729?mt=8"
        let url:NSURL = NSURL(string: iTunelink)!
        
        UIApplication.sharedApplication().openURL(url)
        
        let bundleIdentifier = NSBundle.mainBundle().bundleIdentifier
        print(bundleIdentifier!)
        
    }
    
    @IBAction func link4(sender: UIButton) {
//        print(NSBundle.mainBundle().infoDictionary!["CFBundleName"] as! String)
        UIApplication.sharedApplication().canOpenURL(NSURL(string: "")!)
    }
    
    @IBAction func btn5(sender: AnyObject) {
        let url:NSURL = NSURL(string: "q://")!
        
        UIApplication.sharedApplication().openURL(url)
    }
    
    
}

